### PR TITLE
[2706] Add support for system messages

### PR DESCRIPTION
**Pull request by bychkovdmitry**
https://github.com/GetStream/stream-chat-android/issues/2706

### 🎯 Goal

Add support for system messages

### 🎨 UI Changes

| Before | After |
| --- | --- |
| photo_2021-11-30 21 36 23 | photo_2021-11-30 21 36 20 |

### ☑️ Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] PR targets the `develop` branch
- [x] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [ ] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (docusaurus, tutorial, CMS (task created))
- [x] Reviewers added

### CODE REVIEW

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/viewmodel/messages/MessageListViewModel.kt [hunk 1]
```diff
@@ -559,15 +562,19 @@ public class MessageListViewModel(
...
             }
 
-            groupedMessages.add(
-                MessageItem(
-                    message = message,
-                    groupPosition = position,
-                    parentMessageId = parentMessageId,
-                    isMine = user.id == currentUser?.id,
-                    isInThread = isInThread
+            if (message.isSystem() || message.isError()) {
```
Comment (bychkovdmitry) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/viewmodel/messages/MessageListViewModel.kt:565]: Other SDKs including our UI components treat error and system messages in the same way.

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/messages/list/MessageItem.kt [hunk 2]
```diff
@@ -113,6 +114,14 @@ public fun DefaultMessageItem(
     onLongItemClick: (Message) -> Unit = {},
     onThreadClick: (Message) -> Unit = {},
     onImagePreviewResult: (ImagePreviewResult?) -> Unit = {},
```
Comment (bychkovdmitry) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/messages/list/MessageItem.kt:116]: Probably not the right time, as there is a separate task to introduce date and thread separator slot APIs.

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/messages/list/MessageItem.kt [hunk 3]
```diff
@@ -113,6 +114,14 @@ public fun DefaultMessageItem(
...
+    systemMessageContent: @Composable (SystemMessageItem) -> Unit = {
+        SystemMessage(
```
Comment (skydoves) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/messages/list/MessageItem.kt:118]: Missing KDocs for this new parameter. 😄

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/messages/list/MessageItem.kt [hunk 4]
```diff
@@ -113,6 +114,14 @@ public fun DefaultMessageItem(
...
+            modifier = Modifier
+                .fillMaxWidth()
+                .padding(top = 12.dp, start = 16.dp, end = 16.dp, bottom = 12.dp),
```
Comment (skydoves) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/messages/list/MessageItem.kt:121]: How about using the `vertical` and `horizontal` parameters instead like the below?

```kotlin
.padding(vertical = 12.dp, horizontal = 16.dp)
```

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/messages/list/MessageItem.kt [hunk 5]
```diff
@@ -174,6 +184,26 @@ public fun DefaultMessageItem(
...
+ * @param systemMessageItem The system message item to show.
+ * @param modifier Modifier for styling.
+ */
+@Composable
+public fun SystemMessage(
+    systemMessageItem: SystemMessageItem,
+    modifier: Modifier = Modifier,
+) {
+    Text(
+        modifier = modifier.padding(top = 12.dp, start = 16.dp, end = 16.dp, bottom = 12.dp),
```
Comment (skydoves) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/messages/list/MessageItem.kt:199]: How about using the `vertical` and `horizontal` parameters instead like the below?

```kotlin
.padding(vertical = 12.dp, horizontal = 16.dp)
```

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/util/MessageUtils.kt [hunk 6]
```diff
@@ -30,6 +30,11 @@ internal fun Message.isRegular(): Boolean = type == ModelType.message_regular
  */
 internal fun Message.isSystem(): Boolean = type == ModelType.message_system
 
+/**
+ * @return If the message type is error.
+ */
+public fun Message.isError(): Boolean = type == ModelType.message_error
```
Comment (skydoves) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/util/MessageUtils.kt:36]: Do we need to make this public?

**Review by skydoves — COMMENTED** [review]
Overall looks great to me. 👍 
Left some minor comments.

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/messages/list/MessageItem.kt (hunk 2 above)
Comment (filbabic) [reply stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/messages/list/MessageItem.kt:116]: Yeah, good call - we can do that in a separate task to have everything done in one place! :]

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/state/messages/items/MessageItem.kt [hunk 7]
```diff
@@ -22,6 +22,13 @@ public data class DateSeparator(val date: Date) : MessageListItem()
  */
 public data class ThreadSeparator(val replyCount: Int) : MessageListItem()
 
+/**
+ * Represents a message generated by a system event, such as updating the channel or muting a user.
+ *
+ * @param message The message to show.
+ */
+public data class SystemMessageItem(val message: Message) : MessageListItem()
```
Comment (filbabic) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/state/messages/items/MessageItem.kt:30]: We can rename this to match the new proposal - `SystemMessageState`

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/util/MessageUtils.kt (hunk 6 above)
Comment (filbabic) [reply stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/util/MessageUtils.kt:36]: I would either make it internal, like `isSystem()` or make both of them public!

**Review by filbabic — APPROVED** [review]
Left a couple of comments, but really nice job, thank you so much! :]

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/messages/list/MessageItem.kt (hunk 4 above)
Comment (bychkovdmitry) [reply stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/messages/list/MessageItem.kt:121]: Yep, sorry, was experimenting with the vales and forgot to replace the parameters back.

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/util/MessageUtils.kt (hunk 6 above)
Comment (bychkovdmitry) [reply stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/util/MessageUtils.kt]: Made it internal

**Review by filbabic — APPROVED** [review]
Awesome!

**Pull request merged**